File: resources/browser/three-bridge.js
//-----------------------------------------------------------------------------
// Colorize MultimediaLib
// Copyright 2009-2025 Colorize
// Apache license (http://www.apache.org/licenses/LICENSE-2.0)
//-----------------------------------------------------------------------------

import * as THREE from "three";
import {GLTFLoader} from "three/loaders/GLTFLoader";
import {OBJLoader} from "three/loaders/OBJLoader";
import {MTLLoader} from "three/loaders/MTLLoader";

const FIELD_OF_VIEW = 75;
const NEAR_PLANE = 1;
const FAR_PLANE = 300;

/**
 * Bridge interface for the parts of the Three.js interface implemented in
 * JavaScript. This interface is called from the renderer via TeaVM.
 */
export class ThreeBridge {

    init() {
        this.textureCache = {};

        const width = window.innerWidth;
        const height = window.innerHeight;

        this.scene = new THREE.Scene();
        this.camera = new THREE.PerspectiveCamera(FIELD_OF_VIEW, width / height, NEAR_PLANE, FAR_PLANE);

        this.ambientLight = new THREE.AmbientLight("#DCDCDC", 3);
        this.scene.add(this.ambientLight);

        this.renderer = new THREE.WebGLRenderer();
        this.renderer.setPixelRatio(window.devicePixelRatio);
        this.renderer.setSize(width, height);

        const container = document.getElementById("multimediaLibContainer");
        container.appendChild(this.renderer.domElement);
        window.addEventListener("resize", () => this.resize());
    }

    resize() {
        const width = window.innerWidth;
        const height = window.innerHeight;

        this.camera.aspect = width / height;
        this.camera.updateProjectionMatrix();

        this.renderer.setSize(width, height);
    }

    render() {
        this.renderer.render(this.scene, this.camera);
    }

    changeBackgroundColor(color) {
        this.scene.background = new THREE.Color(color);
    }

    moveCamera(x, y, z, focusX, focusY, focusZ) {
        this.camera.position.set(x, y, z);
        this.camera.lookAt(focusX, focusY, focusZ);
    }

    createBox(width, height, depth, color) {
        const geometry = new THREE.BoxGeometry(width, height, depth);
        const material = this.createMaterial(color, null);
        return new THREE.Mesh(geometry, material);
    }

    createSphere(radius, color) {
        const geometry = new THREE.SphereGeometry(radius);
        const material = this.createMaterial(color, null);
        return new THREE.Mesh(geometry, material);
    }

    createLight(color, intensity) {
        return new THREE.PointLight(color, intensity);
    }

    createMaterial(color, texture) {
        const options = {};
        if (color) {
            options.color = color;
        }
        if (texture) {
            options.map = texture;
            options.transparent = true;
        }
        return new THREE.MeshStandardMaterial(options);
    }

    applyColor(model, color) {
        const material = this.createMaterial(color, null);
        model.traverse(child => {
            if (child.isMesh) {
                child.material = material;
            }
        });
    }

    applyTexture(model, image, x, y, width, height) {
        this.prepareTexture(image, x, y, width, height, texture => {
            model.material = this.createMaterial(null, texture);
            model.needsUpdate = true;
        });
    }

    prepareTexture(image, x, y, width, height, callback) {
        const cacheKey = image.src + "/" + x + "/" + y + "/" + width + "/" + height;

        if (this.textureCache[cacheKey]) {
            callback(this.textureCache[cacheKey]);
            return;
        }

        const canvas = document.createElement("canvas");
        canvas.width = width;
        canvas.height = height;

        const context = canvas.getContext("2d");
        context.drawImage(image, x, y, width, height, 0, 0, width, height);

        const subImage = document.createElement("img");
        subImage.onload = () => {
            const texture = new THREE.Texture(
                subImage,
                THREE.UVMapping,
                THREE.ClampToEdgeWrapping,
                THREE.ClampToEdgeWrapping
            );
            texture.colorSpace = THREE.SRGBColorSpace;
            texture.needsUpdate = true;
            this.textureCache[cacheKey] = texture;
            callback(texture);
        };
        subImage.src = canvas.toDataURL("image/png");
    }

    cloneObject(original) {
        return original.clone();
    }

    loadGLTF(url, callback) {
        const gltfLoader = new GLTFLoader();
        gltfLoader.load(url, gltf => callback(gltf.scene));
    }

    loadOBJ(url, callback) {
        const fileName = url.substring(url.lastIndexOf("/") + 1);
        const dir = url.substring(0, url.lastIndexOf("/") + 1);

        const mtlLoader = new MTLLoader();
        mtlLoader.setPath(dir);
        mtlLoader.load(fileName.replace(".obj", ".mtl"), mtl => {
            mtl.preload();

            const objLoader = new OBJLoader();
            objLoader.setPath(dir);
            objLoader.setMaterials(mtl);
            objLoader.load(fileName, obj => callback(obj));
        });
    }

    project(x, y, z) {
        const position = new THREE.Vector3(x, y, z);
        position.project(this.camera);
        return position;
    }

    castPickRay(pointX, pointY, x0, y0, z0, x1, y1, z1) {
        const raycaster = new THREE.Raycaster();
        raycaster.setFromCamera(new THREE.Vector2(pointX, pointY), this.camera);

        const boundingBox = new THREE.Box3(
            new THREE.Vector3(x0, y0, z0),
            new THREE.Vector3(x1, y1, z1)
        );

        const target = new THREE.Vector3();
        return raycaster.ray.intersectBox(boundingBox, target) != null;
    }
}

File: source/nl/colorize/multimedialib/renderer/teavm/ThreeBridge.java
//-----------------------------------------------------------------------------
// Colorize MultimediaLib
// Copyright 2009-2025 Colorize
// Apache license (http://www.apache.org/licenses/LICENSE-2.0)
//-----------------------------------------------------------------------------

package nl.colorize.multimedialib.renderer.teavm;

import org.teavm.jso.JSFunctor;
import org.teavm.jso.JSObject;
import org.teavm.jso.JSProperty;
import org.teavm.jso.dom.html.HTMLImageElement;

/**
 * TeaVM interface for the {@code three-bridge.js} JavaScript implementation.
 * Includes TeaVM stubs for <a href="https://threejs.org">Three.js</a> classes.
 */
public interface ThreeBridge extends JSObject {

    @JSProperty
    public ThreeObject getScene();

    @JSProperty
    public ThreeLight getAmbientLight();

    public void init();

    public void render();

    public void changeBackgroundColor(String color);

    public void moveCamera(float x, float y, float z, float focusX, float focusY, float focusZ);

    public ThreeObject createBox(float width, float height, float depth, String color);

    public ThreeObject createSphere(float radius, String color);

    public ThreeLight createLight(String color, float intensity);

    public void applyColor(ThreeObject model, String color);

    public void applyTexture(ThreeObject model, HTMLImageElement image, int x, int y, int w, int h);

    public ThreeObject cloneObject(ThreeObject original);

    public void loadGLTF(String url, ModelCallback callback);

    public void loadOBJ(String url, ModelCallback callback);

    public ThreeVector project(float x, float y, float z);

    public boolean castPickRay(float pointX, float pointY,
                               float x0, float y0, float z0, float x1, float y1, float z1);

    /**
     * <a href="https://threejs.org/docs/index.html#api/en/core/Object3D">Object3D</a>
     */
    interface ThreeObject extends JSObject {

        @JSProperty
        public void setVisible(boolean visible);

        @JSProperty
        public ThreeVector getPosition();

        @JSProperty
        public ThreeVector getRotation();

        @JSProperty
        public ThreeVector getScale();

        @JSProperty
        public ThreeMaterial getMaterial();

        public void add(ThreeObject child);

        public void removeFromParent();
    }

    /**
     * <a href="https://threejs.org/docs/index.html#api/en/lights/Light">Light</a>
     */
    interface ThreeLight extends ThreeObject {

        @JSProperty
        public ThreeColor getColor();

        @JSProperty
        public void setIntensity(float intensity);
    }

    /**
     * <a href="https://threejs.org/docs/#api/en/materials/Material">Material</a>
     */
    interface ThreeMaterial extends ThreeObject {

        @JSProperty
        public ThreeColor getColor();
    }

    /**
     * <a href="https://threejs.org/docs/#api/en/textures/Texture">Texture</a>
     */
    interface ThreeTexture extends ThreeObject {
    }

    /**
     * <a href="https://threejs.org/docs/index.html#api/en/math/Vector3">Vector3</a>
     */
    interface ThreeVector extends JSObject {

        @JSProperty
        public float getX();

        @JSProperty
        public float getY();

        @JSProperty
        public float getZ();

        public void set(float x, float y, float z);
    }

    /**
     * <a href="https://threejs.org/docs/#api/en/math/Color">Color</a>
     */
    interface ThreeColor extends ThreeObject {

        public void set(String hex);
    }

    /**
     * <a href="https://threejs.org/docs/#examples/en/loaders/GLTFLoader">GLTFLoader</a>
     */
    @JSFunctor
    @FunctionalInterface
    interface ModelCallback extends JSObject {

        public void onLoadModel(ThreeObject model);
    }
}
